From 0367ca0e54b0b561e9906f52c6d17c96c849bb3c Mon Sep 17 00:00:00 2001
From: yao zhang1 <yao.zhang1@amlogic.com>
Date: Thu, 21 May 2026 09:58:52 +0800
Subject: [PATCH] fastboot: c5 add free rootfs_reserved. [1/1]

PD#SWPL-270624

Problem:
c5 need fastboot

Solution:
c5 add free rootfs_reserved.

Verify:
c5 cb400

Change-Id: If98d23ac96a18f67ab92796626ed296119fc569e
Signed-off-by: yao zhang1 <yao.zhang1@amlogic.com>
---
 init/initramfs.c | 34 ++++++++++++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/init/initramfs.c b/init/initramfs.c
index 1e2955e4a8f8..037462fa5eff 100644
--- a/init/initramfs.c
+++ b/init/initramfs.c
@@ -21,6 +21,7 @@
 
 #ifdef CONFIG_AMLOGIC_FREERTOS_C4
 #include <linux/amlogic/amfc.h>
+#include <linux/of_reserved_mem.h>
 #endif /* CONFIG_AMLOGIC_FREERTOS_C4 */
 
 #include "do_mounts.h"
@@ -727,6 +728,35 @@ struct amfc_bin_hdr {
 };
 
 #define AMFC_BIN_HDR	0x54535a40
+
+static int free_rootfs_memory(void)
+{
+	struct device_node *rootfs_reserved;
+	struct reserved_mem *rmem;
+
+	rootfs_reserved = of_find_compatible_node(NULL, NULL, "amlogic, aml_rootfs_memory");
+	if (!rootfs_reserved) {
+		pr_err("rootfs_reserved device node find error\n");
+		return 1;
+	}
+
+	rmem = of_reserved_mem_lookup(rootfs_reserved);
+	if (!rmem) {
+		pr_err("%s no such reserved mem\n", (char *)rootfs_reserved->name);
+		return 1;
+	}
+
+	if (!rmem->base || !rmem->size) {
+		pr_err("%s unexpected reserved memory\n", (char *)rootfs_reserved->name);
+		return 1;
+	}
+
+	free_reserved_area(__va(rmem->base),
+			   __va(PAGE_ALIGN(rmem->base + rmem->size)),
+			   0,
+			   "free_aml_rootfs_memory");
+	return 0;
+}
 #endif /* CONFIG_AMLOGIC_FREERTOS_C4 */
 
 #ifdef CONFIG_AMLOGIC_FREERTOS_C4
@@ -826,6 +856,10 @@ static void __init do_populate_rootfs(void *unused, async_cookie_t cookie)
 	initrd_end = 0;
 
 	init_flush_fput();
+#ifdef CONFIG_AMLOGIC_FREERTOS_C4
+	if (free_rootfs_memory())
+		pr_err("Failed to free rootfs memory\n");
+#endif
 }
 
 #ifndef CONFIG_AMLOGIC_FREERTOS_C4
-- 
2.46.0

